File: api/BeSureApi/Models/ContractBankGuarantee.cs
namespace BeSureApi.Models
{
    public class BankGuaranteeDetails
    {
        public int ContractId { get; set; }
        public int GuaranteeType { get; set; }
        public string GuaranteeNumber { get; set; }
        public int BankBranchInfoId { get; set; }
        public string GuaranteeStartDate { get; set; }
        public string GuaranteeEndDate { get; set; }
        public Decimal GuaranteeAmount { get; set; }
        public string Remarks { get; set; }
        public int GuaranteeClaimPeriodInDays { get; set; }
    }
    public class BankGuaranteeUpdateDetails
    {
        public int Id { get; set; }
        public int GuaranteeType { get; set; }
        public string GuaranteeNumber { get; set; }
        public int BankBranchInfoId { get; set; }
        public string GuaranteeStartDate { get; set; }
        public string GuaranteeEndDate { get; set; }
        public Decimal GuaranteeAmount { get; set; }
        public string Remarks { get; set; }
        public int GuaranteeClaimPeriodInDays { get; set; }
    }
    public class BankGuaranteeList
    {
        public int Id { get; set; }
        public int GuaranteeStatusId { get; set; }
        public string GuaranteeType { get; set; }
        public string GuaranteeStatus { get; set; }
        public string GuaranteeNumber { get; set; }
        public string BranchName { get; set; }
        public DateTime GuaranteeEndDate { get; set; }
        public Decimal GuaranteeAmount { get; set; }
        public string Remarks { get; set; }
        public int GuaranteeClaimPeriodInDays { get; set; }
        public string CreatedBy { get; set; }
        public DateTime CreatedOn { get; set; }
    }
}

File: ui/src/types/contractBankGuarantee.ts
import { Decoder, array, boolean, nullable, number, object, string } from "decoders";

export interface BankGuaranteeCreate {
    ContractId: string;
    GuaranteeType: number;
    GuaranteeNumber: string;
    BankBranchInfoId: number;
    GuaranteeStartDate: string;
    GuaranteeEndDate: string;
    GuaranteeAmount: number;
    Remarks: string;
    GuaranteeClaimPeriodInDays: number;
}

export interface BankGuaranteeCreateResult {
    IsBankGuaranteeCreated: Boolean;
}

export const bankGuaranteeCreateResultDecoder: Decoder<BankGuaranteeCreateResult> = object({
    IsBankGuaranteeCreated: boolean,
});

export interface BankGuaranteeDetails {
    Id: number;
    GuaranteeStatusId: number;
    GuaranteeStatus: string;
    GuaranteeType: string;
    GuaranteeNumber: string;
    BranchName: string;
    GuaranteeEndDate: string;
    GuaranteeAmount: number;
    Remarks: string;
    GuaranteeClaimPeriodInDays: number;
    CreatedBy: string;
    CreatedOn: string;
}

export interface BankGuaranteeList {
    BankGuarantees: BankGuaranteeDetails[]
}

export const bankGuaranteeDetailsDecoder: Decoder<BankGuaranteeDetails> = object({
    Id: number,
    GuaranteeStatusId: number,
    GuaranteeStatus: string,
    GuaranteeType: string,
    GuaranteeNumber: string,
    BranchName: string,
    GuaranteeEndDate: string,
    GuaranteeAmount: number,
    Remarks: string,
    GuaranteeClaimPeriodInDays: number,
    CreatedBy: string,
    CreatedOn: string,
})

export const bankGuaranteeListDecoder: Decoder<BankGuaranteeList> = object({
    BankGuarantees: array(bankGuaranteeDetailsDecoder),
});

export interface BankGuaranteeEditDetail {
    Id: number;
    GuaranteeType: number;
    GuaranteeNumber: string;
    BankBranchInfoId: number;
    GuaranteeStartDate: string;
    GuaranteeEndDate: string;
    GuaranteeAmount: number;
    Remarks: string;
    GuaranteeClaimPeriodInDays: number;
}

export const bankGuaranteeEditDetailDecoder: Decoder<BankGuaranteeEditDetail> = object({
    Id: number,
    GuaranteeType: number,
    GuaranteeNumber: string,
    BankBranchInfoId: number,
    GuaranteeStartDate: string,
    GuaranteeEndDate: string,
    GuaranteeAmount: number,
    Remarks: string,
    GuaranteeClaimPeriodInDays: number,
});

export interface BankGuaranteeEditDetails {
    BankGuaranteeDetails: BankGuaranteeEditDetail;
}

export const bankGuaranteeEditDetailsDecoder: Decoder<BankGuaranteeEditDetails> = object({
    BankGuaranteeDetails: bankGuaranteeEditDetailDecoder,
});

export interface BankGuaranteeEditResponse {
    IsBankGuaranteeUpdated: Boolean;
}

export const bankGuaranteeEditDecoder: Decoder<BankGuaranteeEditResponse> = object({
    IsBankGuaranteeUpdated: boolean,
});

export interface Select {
    value: any,
    label: any
}

export interface SelectDetails {
    Select: Select[]
}

export interface Configurations {
    GuaranteeType: Select[],
    BankBranchNames: Select[],
}
